File: Easy_Problems/1. Two Number Sum/Java/Solution3.java
import java.util.HashMap;
import java.util.Scanner;
import java.util.Map;

class Solution3{

    // O(nlog(n)) | O(1) space
    // Pointer Approach (Sort + Two Pointers)
    private static int[] findTwoSum_3(int[] nums, int target) {
        Arrays.sort(nums);
        int left = 0;
        int right = nums.length - 1;
        while(left < right) {
            if(nums[left] + nums[right] == target) {
                return new int[] {nums[left], nums[right]};
            } else if (nums[left] + nums[right] < target) {
                left++;
            } else {
                right--;
            }
        }
        return new int[] {};
    }


    public static void main(String[] args) {
        Scanner sc = new Scanner(System.in);

        int n = sc.nextInt();
        int[] nums = new int[n];

        for(int i = 0; i < n; i++) {
            nums[i] = sc.nextInt();
        }
        int target = sc.nextInt();

        sc.close();

        int[] pair_arr = findTwoSum_3(nums, target);

        if (pair_arr.length == 2) {
            System.out.println(pair_arr[0] + " " + pair_arr[1]);
        } else {
            System.out.println("No pair found!");
        }
    }
}

File: Easy_Problems/1. Two Number Sum/Javascript/Solution3.js
// O(nlog(n)) | O(1) space
// Pointer Approach (Sort + Two Pointers)

function twoNumberSum(array, targetSum) {
    array.sort((a, b) => a - b);
    let left = 0;
    let right = array.length - 1;
    while (left < right) {
        const currentSum = array[left] + array[right];
        if (currentSum === targetSum) {
            return [array[left], array[right]];
        } else if (currentSum < targetSum) {
            left++;
        } else if (currentSum > targetSum) {
            right--;
        }
    }
    return [];
}

twoNumberSumTest = () => {
    console.log(twoNumberSum([3, 5, -4, 8, 11, 1, -1, 6], 10)); // [-1, 11]
    console.log(twoNumberSum([1, 2, 3, 4, 5, 6, 7, 8, 9, 10], 15)); // [10, 5]
    console.log(twoNumberSum([1, 3, 5, 7, 9], 8)); // [3, 5]
    console.log(twoNumberSum([1, 2, 4, 6, 8, 10, 12, 14, 16, 18], 17)); // [1, 16]
    console.log(twoNumberSum([3, 5, 7, 9], -8)); // []
    console.log(twoNumberSum([3, 5, 7, 9], 0)); // []
    console.log(twoNumberSum([3, 5, 7, 9], 1)); // []
    console.log(twoNumberSum([3, 5, 7, 9], 2)); // []
};
